Sites Management Page: Improve render performance by paginating when greater than 96 sites

danielbachhuber opened the pull request:
Fixes #67612

## Proposed Changes

* Applies pagination when displaying greater than 96 sites in order to improve render performance. Uses `96` instead of `100` as the default `perPage` to better support 3 and 2 column rows.
* Exports `handleQueryParamChange` function for use by `<SitesDashboard />` component.

<img width="1347" alt="image" src="https://user-images.githubusercontent.com/36432/189738393-ff4dd278-b164-4cf9-9b37-e42d06829f8d.png">

## Testing Instructions

1. Navigate to `/sites` with a WordPress user that has greater than 3 sites.
2. Add `?per-page=3` to the URL.
3. Observe the pagination appear at the bottom of the sites list.
4. Click on the pagination component buttons and observe expected behavior (paginating through the results).

One note: when manipulating `?per-page=3`, you may end up in this stacked component scenario:

<img width="1378" alt="image" src="https://user-images.githubusercontent.com/36432/189739015-86924661-f7af-4aa3-a8d0-ab024d4bfb30.png">

While stacking the pagination and hidden sites components doesn't look like the optimal design, I don't think it's really worth solving for because it's a really uncommon scenario.

danielbachhuber committed:
Move function and export for use in `SitesDashboard`

danielbachhuber committed:
Apply pagination when greater than 96 sites displayed

danielbachhuber committed:
Merge branch 'trunk' into add/sites-management-page-pagination

wojtekn commented on client/sites-dashboard/controller.tsx:
```diff
@@ -68,6 +68,10 @@ export function sitesDashboard( context: PageJSContext, next: () => void ) {
 			<SitesDashboard
 				queryParams={ {
+					page: context.query.page ? parseInt( context.query.page ) : undefined,
+					perPage: context.query[ 'per-page' ]
```
Do we want to allow users to set it to any value? Or should we accept the value from the parameter only if it's lower or equal to 96?

danielbachhuber replied:
> Do we want to allow users to set it to any value? Or should we accept the value from the parameter only if it's lower or equal to 96?

Given the value only modifies displayed state, and isn't passed through to an API call, I don't think we need to do any more processing than we already are.

wojtekn reviewed (changes requested):
@danielbachhuber looks good and works well for me. That feature should help with handling that edge case.

I've just noticed behavior that we may need to adjust - when users are on the second page and start searching for a phrase that can be found on the first page, they will see “No sites match your search”, and they will lose a way to navigate back to the first page. The same happens for the status filter.

Should we reset the page number when users start searching?

There is also a minor styling issue on mobile view - utility buttons partially cover the pagination. We may need to add some more space below the paginator component:

![Screen Shot 2022-09-13 at 10 18 56](https://user-images.githubusercontent.com/727413/189849894-30f9826c-3271-430a-8485-b97090482250.png)

danielbachhuber committed:
Merge branch 'trunk' into add/sites-management-page-pagination

danielbachhuber committed:
Reset `page` query parameter after initiating search or filter

danielbachhuber committed:
Add more bottom padding to avoid overlap with page tools

danielbachhuber committed:
Another piece to e8a549f28dbab6794a2c453c0b70187b215434a4

danielbachhuber committed:
Clean up type issues

danielbachhuber committed:
Avoid errant 'No search results found' state while preferences load

danielbachhuber commented:
> I've just noticed behavior that we may need to adjust - when users are on the second page and start searching for a phrase that can be found on the first page, they will see “No sites match your search”, and they will lose a way to navigate back to the first page. The same happens for the status filter.
> 
> Should we reset the page number when users start searching?

@wojtekn Great find! I've updated the logic to reset `page` when a search or filter is initiated: 1699d845ad62bca50b68e36ec475757dc2530418

> There is also a minor styling issue on mobile view - utility buttons partially cover the pagination. We may need to add some more space below the paginator component:

Cleaned up with e8a549f28dbab6794a2c453c0b70187b215434a4 and c67ba827bd840ec71e8298ebe197053b6a2705cb

danielbachhuber committed:
Use `undefined` to fix type issues

wojtekn reviewed (approved):
@danielbachhuber works great now!
